nifi-nar-bundles/nifi-cdc/nifi-cdc-postgresql-bundle/nifi-cdc-postgresql-processors/src/main/java/org/apache/nifi/cdc/postgresql/processors/CaptureChangePostgreSQL.java

```diff
@@ -0,0 +1,468 @@
+            this.connectionManager.createSQLConnection();
+            this.connectionManager.createReplicationConnection();
+
+            stateMap = stateManager.getState(Scope.CLUSTER);
+
+            this.pgEasyReplication = new PGEasyReplication(publicationName, replicationSlotName, this.connectionManager);
+            this.pgEasyReplication.initializeLogicalReplication(dropSlotIfExists);
+
+            if (stateMap.get(LAST_LSN_RECEIVED) != null) {
+                this.lastLSNReceived = Long.parseLong(stateMap.get(LAST_LSN_RECEIVED));
+                this.hasRun.set(true);
+            } else {
+                this.hasRun.set(false);
+            }
+
+        } catch (final Exception ioe) {
+            logger.error("Failed to retrieve values from the State Manager. Will not attempt " + "connection until this is accomplished.", ioe);
+            context.yield();
+            throw new ProcessException(ioe.getMessage(), ioe);
+        }
+    }
+
+    List<Event> getData() {
+
+        List<Event> events = new ArrayList<>();
+
+        if (!this.hasRun.get() && this.hasSnapshot) {
+            events.add(this.pgEasyReplication.getSnapshot());
+        }
+
+        if (!this.hasRun.get() && this.initialLSN != null) {
+            events.add(this.pgEasyReplication.readEvent(true, this.includeBeginCommit, MIME_TYPE_VALUE, initialLSN));
+        }
+
+        events.add(this.pgEasyReplication.readEvent(true, this.includeBeginCommit));
+
+        if (events != null)
+            this.hasRun.set(true);
+
+        return events;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final ComponentLog logger = getLogger();
+
+        // Indicate that this processor has executed at least once, so we know whether
+        // or not the state values are valid and should be updated
+        if (this.pgEasyReplication == null) {
+            setup(context);
+        }
+
+        List<FlowFile> listFlowFiles = new ArrayList<>();
+
+        List<Event> events = getData();
+
+        if (events == null) {
+            logger.error("Failed to retrieve events.");
+            context.yield();
+            throw new ProcessException();
+        }
+
+        for (Event event : events) {
+            if (event != null) {
+                LinkedList<String> listData = event.getData();
+
+                if (!listData.isEmpty()) {
+                    for (String data : listData) {
+                        FlowFile flowFile = null;
+                        flowFile = session.create();
+
+                        flowFile = session.write(flowFile, new OutputStreamCallback() {
+                            @Override
+                            public void process(final OutputStream out) throws IOException {
+                                out.write(data.getBytes(StandardCharsets.UTF_8));
```

Comment:
Related to my comment above about multiple event records per flow file, would it also be better to allow the output in any format supported by NiFi RecordSetWriters? Again the existing CDC processor used JSON because the record capability wasn't present, but since then the conventional wisdom is to put larger number of records in each flow file and allow them to be written using a configured RecordSetWriter.

This is a big change to the existing PR, so if you'd like it to get into the codebase as-is and then add the record stuff later, that's fine with me. Let's just get a Jira written (depending on NIFI-4239) that describes the additional capability to be added. IMO this would be adding all available events (or up to a configurable limit) and allowing the user to select a RecordSetWriter for output. If no writer is configured (and none would be by default, it would be an optional property), the behavior would fall back to the current behavior, writing 1-3 events as JSON into the flow file.  What are your thoughts?

Reply:
Yeah, we must use record capability. That can improve this processor in several aspects. As you suggested, we will proceed with opening a new Jira for this .